README.md
range_repair.py
======================

A Python script to repair the primary range of a Cassandra node in N discrete steps [using best practices](http://www.datastax.com/dev/blog/advanced-repair-techniques).

### Background
When Cassandra begins the repair process it constructs a [merkle tree](http://en.wikipedia.org/wiki/Merkle_tree), which is a tree of hashes over segments of data that the node is responsible for. The node compares it's tree to that of the replicas, if there is a difference in the hash values for any of the nodes then the segment for that hash is requested from the replica and is re-inserted.

By default the Merkle tree for Cassandra represents 15 discreet segments, which means that the data on the node is broken into 15 pieces. If the data for one of these 15 pieces is different from that of the replicas then it will result in 1/15th of the data being streamed and re-inserted. This can cause problems for two use cases: dense nodes, and DSE Solr nodes running at or near capacity.

#### Solr Nodes
For DSE Solr nodes when the data transferred as the result of a repair is re-inserted it is also re-indexed because each host maintains it's own independent set of indexes. If the node is already at or near capacity then the additional strain caused by the repair/re-index can push it over the edge. As a last result Cassandra will shed load by dropping mutations. If a mutation is dropped the data will at some point need to be brought over from the replicas through the repair process which unfortunately begins a never ending cycle of re-index/repair.

#### Dense Nodes
For clusters that have a large amount of data per node the repair process could require an unacceptably large amount of data to be streamed and re-inserted.

### How the script works
The script works by figuring out the primary range for the node that it's being executed on, and instead of running repair on the entire range, run the repair on only a smaller sub-range. When a repair is initiated on a sub-range Cassandra constructs a merkle tree only for the range specified, which in turn divides the much smaller range into 15 segments. If there is disagreement in any of the hash values then a much smaller portion of data needs to be transferred which lessens load on the system.

### Multiple Datacenters
If you have multiple datacenters in your ring, then you MUST specify the name of the datacenter containing the node you are repairing as part of the command-line options (--datacenter=DCNAME).  Failure to do so will result in only a subset of your data being repaired (approximately data/number-of-datacenters).  This is because nodetool has no way to determine the relevant DC on its own, which in turn means it will use the tokens from every ring member in every datacenter.

### Options

```
Usage: range_repair.py [options]

Options:
  -h, --help            show this help message and exit
  -k KEYSPACE, --keyspace=KEYSPACE
                        Keyspace to repair (REQUIRED)
  -c COLUMNFAMILY, --columnfamily=COLUMNFAMILY
                        ColumnFamily to repair, can appear multiple times
  -H HOST, --host=HOST  Hostname to repair [default: $HOSTNAME]
  -s STEPS, --steps=STEPS
                        Number of discrete ranges [default: 100]
  -o OFFSET, --offset=OFFSET
                        Number of tokens to skip [default: 0]   
  -n NODETOOL, --nodetool=NODETOOL
                        Path to nodetool [default: nodetool]
  -w WORKERS, --workers=WORKERS
                        Number of workers to use for parallelism [default: 1]
  -D DATACENTER, --datacenter=DATACENTER
  -l, --local           Restrict repair to the local DC
  -p, --par             Carry out a parallel repair (post-2.x only)
  -i, --inc             Carry out an incremental repair (post-2.1 only).
  -S, --snapshot        Use snapshots (pre-2.x only)
  -v, --verbose         Verbose output
  -d, --debug           Debugging output
  --dry-run             Do not execute repairs.
  --syslog=FACILITY     Send log messages to the syslog
  --logfile=FILENAME    Send log messages to a file
```

### Sample

```
$ ./range_repair.py -k demo_keyspace
INFO       2014-05-09 17:31:33,503    get_ring_tokens                 66  : running nodetool ring, this will take a little bit of time
DEBUG      2014-05-09 17:31:39,057    get_ring_tokens                 72  : ring tokens found, creating ring token list...
DEBUG      2014-05-09 17:31:40,207    get_host_tokens                 86  : host tokens found, creating host token list...
DEBUG      2014-05-09 17:31:40,208    repair_keyspace                 170 : repair over range (-2974082934175371230, -2971948823734978979] with 100 steps for keyspace demo_keyspace
DEBUG      2014-05-09 17:31:40,208    repair_keyspace                 176 : step 0100 repairing range (-2974082934175371230, -2974061593070967308] for keyspace demo_keyspace ...
DEBUG      2014-05-09 17:32:47,508    repair_keyspace                 182 : SUCCESS
DEBUG      2014-05-09 17:32:47,509    repair_keyspace                 176 : step 0099 repairing range (-2974061593070967308, -2974040251966563386] for keyspace demo_keyspace ...
DEBUG      2014-05-09 17:33:54,904    repair_keyspace                 182 : SUCCESS
...
```

### Dependencies
-   Python 2.6
-   Cassandra ```nodetool``` must exist in the ```PATH```

## How To Test/Contribute
1. You will need pbr (https://pypi.python.org/pypi/pbr) available in your $PYTHONPATH
1. Go to the root of the cassandra_range_repair project on your local Git clone.
2. Make the necessary changes you want included
3. If you are testing the package on Unix/Linux based environment, make sure the package builds and tests ok

```
make clean
make build
make test
make debian
```

4. Create a pull request to merge to the main branch
5. Be patient and 

### History
- Originally by [Matt Stump](https://github.com/mstump)
- Converted to work with vnodes by [Brian Gallew](https://github.com/BrianGallew)
- Additional functionality by [Eric Lubow](http://github.com/elubow)
- Support for multiprocessing performed by [Brian Gallew](https://github.com/BrianGallew) with credit to [M. Jaszczyk](https://github.com/mjaszczyk)
- Multiple datacenter support by [Brian Gallew](https://github.com/BrianGallew)
- Support debian packaging [Venkatesh Kaushik] (https://github.com/higgsmass)



ChangeLog
CHANGES
=======

* Update range_repair.py
* minor modifications to requirements.txt
* changes for debian packaging of cassandra_range_repair
* Removing redundant local flag (part 2 of fix to use local and pr flags mutually exclusively)
* Fix - use local or pr flag (cannot use both)
* Add support for incremental repairs (post 2.1)
* Adding token offset argument
* Updating readme (removing reference to LOG_LEVEL env var and documenting some new flags)
* fixed node range assignment (each node is responsible for the region of the ring between itself and its predecessor)
* added support to parallel repairs (post 2.x)
* Add site to the docstring
* While working on #14 I discovered that not only was there a larger than expected number of repairs being performed, but also that there seemed to be one repair for every token where the start point was larger than the endpoint. This led me to refactor the entire sub-range generation process
* fix duplicated log messages
* repair() doesn't return anything, so checking the return value is silly
* Make --dry-run print out the step information as well as the actual nodetool command to be run
* Missed checking for the keyspace param in the repair_range(
* Add more logging options (syslog and logfile). Enable a "dry-run" mode that doesn't execute the repairs. Don't require a keyspace to be specified unless one or more columfamilies are specified
* I don't think the repairs ever worked for negative murmur3 tokens, because the old format string was only ever producing a 20-byte string *including* the minus sign. The new format string always prints the sign, and that work for 1.2.x. Someone testing on 2.x would be good
* Add a CLI flag to specify the datacenter
* Don't pre-join the arguments to run_command
* Removed old partitioning constants. Decremented Murmur3 max value by one to match documentation
* removed a debugging list reducer
* Added default values to help output. Swapped out range() for xrange() because xrange() can't handle such large values
* Updated README
* This *works*, but still doesn't have multiprocessing
* First major sweep through, still doesn't actually have any multiprocessing bits yet
* The os module is no longer required
* Added verbose/debug flags, don't bother with the environment variables anymore
* restore help argument
* fix docstring argument order
* spelling failure
* add host support to nodetool ring command
* update readme
* filter joining nodes
* remove stray import
* show progress with logging.info
* Add host argument
* https://github.com/mstump/cassandra_range_repair/issues/2
* better completion logging
* updated README
* merging code cleanup
* code cleanup and better logging
* find() should return -1 if nothing is found
* Extend it to work with vnodes. NB: this has not been tested against an old-fashioned, non-vnode ring
* added lots of explanation to the README about the problems the script attempts to solve and how
* made the script python2.5 compatible, and tested it
* also use the -local option when performing the repair
* also use the -snapshot option when performing the repair
* Update README.md
* Update README.md
* put the __main__ trick in place
* Create README.md
* Update LICENSE.md
* Create LICENSE.md
* initial commit


tests/test_ranges.py
#! /usr/bin/env python

from __future__ import print_function
import os, sys, unittest, pkg_resources, mock, logging
sys.path.insert(0, '..')
sys.path.insert(0, '.')

sys.path.insert(0,os.path.abspath(__file__+"/../../src"))

import range_repair

def fake_init(self, options):
    '''Initialize the Token Container by getting the host and ring tokens and
    then confirming the values used for formatting and range
    management.
    :param options: OptionParser result
    :returns: None
    '''
    self.options = options
    self.host_tokens = []
    self.ring_tokens = []
    self.host_token_count = -1
    return

class FakeOptions: pass
    
class range_tests(unittest.TestCase):
    def setUp(self):
        range_repair.Token_Container.__init__ = fake_init
        f = FakeOptions()
        f.keyspace='pathdb'
        f.columnfamily='path_claims'
        f.host='db-cdev-1.phx3.llnw.net'
        f.steps=5
        f.nodetool='nodetool'
        f.workers=1
        f.local=''
        f.snapshot=''
        f.verbose=True
        f.debug=True
        self.f = f
        return

    def test_Murmur3_range_start_zero(self):
        resultset = []
        t = range_repair.Token_Container(self.f)
        for x in t.sub_range_generator(0, 3000, steps=5):
            resultset.append(x[0])
        self.assertEquals(resultset, [t.format(0), t.format(600), t.format(1200), t.format(1800), t.format(2400)])
        return

    def test_Murmur3_range_end_zero(self):
        resultset = []
        t = range_repair.Token_Container(self.f)
        for x in t.sub_range_generator(-3000, 0, steps=5):
            resultset.append(x[0])
        self.assertEquals(resultset, [t.format(-3000), t.format(-2400), t.format(-1800), t.format(-1200), t.format(-600)])
        return

    def test_Murmur3_range_wrap(self):
        resultset = []
        t = range_repair.Token_Container(self.f)
        endpoint = (2**63)-30
        for x in t.sub_range_generator(endpoint, -endpoint, steps=6):
            resultset.append(x[0])
        self.assertEquals(len(resultset), 7)
        return
        
    def test_Random_range_start_zero(self):
        resultset = []
        t = range_repair.Token_Container(self.f)
        t.ring_tokens.append(0)
        t.check_for_MD5_tokens()
        for x in t.sub_range_generator(0, 3000, steps=5):
            resultset.append(x[0])
        self.assertEquals(resultset, [t.format(0), t.format(600), t.format(1200), t.format(1800), t.format(2400)])
        return

    def test_Random_range_end_zero(self):
        resultset = []
        t = range_repair.Token_Container(self.f)
        t.ring_tokens.append(0)
        t.check_for_MD5_tokens()
        for x in t.sub_range_generator(-3000, 0, steps=5):
            resultset.append(x[0])
        self.assertEquals(resultset, [t.format(-3000), t.format(-2400), t.format(-1800), t.format(-1200), t.format(-600)])
        return

    def test_Random_range_wrap(self):
        resultset = []
        t = range_repair.Token_Container(self.f)
        endpoint = (2**63)-30
        t.ring_tokens.append(0)
        t.check_for_MD5_tokens()
        for x in t.sub_range_generator(endpoint, -endpoint, steps=6):
            resultset.append(x[0])
        self.assertEquals(len(resultset), 6)
        return
    def test_Murmur3_format_length(self):
        t = range_repair.Token_Container(self.f)
        self.assertEquals(21, len(t.format(0)))
        self.assertEquals(21, len(t.format(100)))
        self.assertEquals(21, len(t.format(-100)))



tests/rangerepair_test.py
#! /usr/bin/env python
import sys
import os

sys.path.insert(0, '..')
sys.path.insert(0, '.')

sys.path.insert(0,os.path.abspath(__file__+"/../../src"))

import range_repair


if __name__ == '__main__':
    suite1 = unittest.TestLoader().loadTestsFromTestCase(execution_count_tests)
    unittest.TextTestRunner(verbosity=2).run(suite1)
    suite2 = unittest.TestLoader().loadTestsFromTestCase(range_tests)
    unittest.TextTestRunner(verbosity=2).run(suite2)


src/range_repair.py
#!/usr/bin/env python
"""
This script will allow for smaller repairs of Cassandra ranges.

See the tests subdirectory for example code.

Source: https://github.com/BrianGallew/cassandra_range_repair
"""
from optparse import OptionParser

import logging
import subprocess
import sys
import multiprocessing
import platform
import re

class Token_Container:
    RANGE_MIN = -(2**63)
    RANGE_MAX = (2**63)-1
    FORMAT_TEMPLATE = "{0:+021d}"
    def __init__(self, options):
        '''Initialize the Token Container by getting the host and ring tokens and
        then confirming the values used for formatting and range
        management.
        :param options: OptionParser result
        :returns: None
        '''
        self.options = options
        self.local_nodes = []
        self.host_tokens = []
        self.ring_tokens = []
        self.host_token_count = -1
        self.get_local_nodes()
        self.get_host_tokens()
        self.get_ring_tokens()
        self.check_for_MD5_tokens()
        return

    def get_local_nodes(self):
        '''In a multi-DC environment, it is important to *only* consider tokens on
        members of the local ring.

        '''
        if not self.options.datacenter:
            logging.debug("No datacenter specified, all ring members' tokens will be considered")
            return
        logging.debug("Determining local ring members")
        cmd = [self.options.nodetool, "-h", self.options.host, "gossipinfo"]
        success, _, stdout, stderr = run_command(*cmd)

        if not success:
            raise Exception("Died in get_ring_tokens because: " + stderr)

        # This is a really well-specified value.  If the format of the
        # output of 'nodetool gossipinfo' changes, this will have to be
        # revisited.
        search_regex = "DC:\d:{datacenter}".format(datacenter=self.options.datacenter)
        for paragraph in stdout.split("/"):
            if not re.search(search_regex, paragraph):
                continue
            self.local_nodes.append(paragraph.split()[0])
        logging.info("Local nodes: " + " ".join(self.local_nodes))
        return

    def check_for_MD5_tokens(self):
        """By default, the Token_Container assumes that the Murmur3 partitioner is
        in use.  If that's true, then the first token in the ring should
        have a negative value as long as the cluster has at least 3
        (v)nodes.  If the first token is not negative, switch the class
        constants for the values associated with Random paritioner.
        :returns: None
        """
        if not self.ring_tokens[0] < 0:
            self.FORMAT_TEMPLATE = "{0:039d}"
            self.RANGE_MIN = 0
            self.RANGE_MAX = (2**127) - 1
        return

    def get_ring_tokens(self):
        """Gets the token information for the ring
        :returns: None
        """
        logging.info("running nodetool ring, this will take a little bit of time")
        cmd = [self.options.nodetool, "-h", self.options.host, "ring"]
        success, _, stdout, stderr = run_command(*cmd)

        if not success:
            raise Exception("Died in get_ring_tokens because: " + stderr)

        logging.debug("ring tokens found, creating ring token list...")
        for line in stdout.split("\n")[4:]:
            segments = line.split()
            # Filter tokens from joining nodes
            if (len(segments) != 8) or (segments[3] == "Joining"):
                logging.debug("Discarding: %s", line)
                continue
            # If a datacenter has been specified, filter nodes that are in
            # different datacenters.
            if self.options.datacenter and not segments[0] in self.local_nodes:
                logging.debug("Discarding node/token %s/%s", segments[0], segments[-1])
                continue
            self.ring_tokens.append(long(segments[-1]))
            logging.debug(str(self.ring_tokens))
        self.ring_tokens.sort()
        logging.info("Found {0} tokens".format(len(self.ring_tokens)))
        logging.debug(self.ring_tokens)
        return

    def get_host_tokens(self):
        """Gets the tokens ranges for the target host
        :returns: None
        """
        cmd = [self.options.nodetool, "-h", self.options.host, "info", "-T"]
        success, _, stdout, stderr = run_command(*cmd)
        if not success or stdout.find("Token") == -1:
            logging.error(stdout)
            raise Exception("Died in get_host_tokens, success: %d, stderr: %s" % (success, stderr))

        for line in stdout.split("\n"):
            if not line.startswith("Token"): continue
            parts = line.split()
            self.host_tokens.append(long(parts[-1]))
        self.host_tokens.sort()
        self.host_token_count = len(self.host_tokens)
        logging.debug("%d host tokens found", self.host_token_count)
        return


    def format(self, value):
        '''Return the correctly zero-padded string for the token.
        :returns: the properly-formatted token.
        '''
        return self.FORMAT_TEMPLATE.format(value)

    def get_preceding_token(self, token):
        """get the end token of the previous range
        :param token: Reference token
        :returns: The token that falls immediately before the argument token
        """
        for i in reversed(self.ring_tokens):
            if token > i:
                return i
        # token is the smallest value in the ring.  Since the rings wrap around,
        # return the last value.
        return self.ring_tokens[-1]

    def sub_range_generator(self, start, stop, steps=100):
        """Generate $step subranges between $start and $stop
        :param start: beginning token in the range
        :param stop: first token of the next range
        :param steps: number of sub-ranges to create
        :returns: string-formatted start value, string-formatted end value, current step number
        There is special-case handling for when there are more steps than there
        are keys in the range: just return the start and stop values.
        """
        step = 0
        # This first case works for all but the highest-valued token.
        if stop > start:
            if start+steps < stop+1:
                step_increment = ((stop - start) / steps)
                # We would have an extra, tiny step in the event the range
                # is not evenly divisible by the number of steps.  This may
                # give us one larger step at the end.
                step_list = [self.format(x) for x in range(start, stop, step_increment)][0:steps]
            else:
                step += 1
                yield self.format(start), self.format(stop), step
        else:                     # This is the wrap-around case
            distance = (self.RANGE_MAX - start) + (stop - self.RANGE_MIN)
            if distance > steps-1:
                step_increment = distance / steps
                # Can't use xrange here because the numbers are too large!
                step_list = [self.format(x) for x in range(start, self.RANGE_MAX, step_increment)]
                step_list.extend([self.format(x) for x in range(self.RANGE_MIN, stop, step_increment)])
                if len(step_list) > steps-1:
                    step_list.pop()
            else:
                step += 1
                yield self.format(start), self.format(stop), step
        step_list.append(self.format(stop)) # Add the final number to the list
        # Now iterate pair-wise over the list
        while len(step_list) > 1:
            step += 1
            yield step_list[0], step_list[1], step
            step_list.pop(0)

def run_command(*command):
    """Execute a shell command and return the output
    :param command: the command to be run and all of the arguments
    :returns: success_boolean, command_string, stdout, stderr
    """
    cmd = " ".join(command)
    logging.debug("run_command: " + cmd)
    proc = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = proc.communicate()
    return proc.returncode == 0, cmd, stdout, stderr

def repair_range(options, start, end, step, nodeposition):
    """Repair a keyspace/columnfamily between a given token range with nodetool
    :param options: OptionParser result
    :param start: Beginning token in the range to repair (formatted string)
    :param end: Ending token in the range to repair (formatted string)
    :param step: The step we're executing (for logging purposes)
    :param nodeposition: string to indicate which node this particular step is for.
    :returns: None
    """
    logging.debug(
        "{nodeposition} step {step:04d} repairing range ({start}, {end}) for keyspace {keyspace}".format(
            step=step,
            start=start,
            end=end,
            nodeposition=nodeposition,
            keyspace=options.keyspace or "<all>"))

    cmd = [options.nodetool, "-h", options.host, "repair"]
    if options.keyspace: cmd.append(options.keyspace)
    cmd.extend(options.columnfamily)

    # -local flag cannot be used in conjunction with -pr
    if options.local:
        cmd.extend([options.local])
    else:
        cmd.extend(["-pr"])
        
    cmd.extend([options.par, options.inc, options.snapshot,
                 "-st", start, "-et", end])

    if not options.dry_run:
        success, cmd, _, stderr = run_command(*cmd)
    else:
        print "{step:04d}/{nodeposition}".format(nodeposition=nodeposition, step=step), " ".join(cmd)
        success = True

    if not success:
        logging.error("FAILED: {nodeposition} step {step:04d} {cmd}".format(nodeposition=nodeposition, step=step, cmd=cmd))
        logging.error(stderr)
        return
    logging.debug("{nodeposition} step {step:04d} complete".format(nodeposition=nodeposition,step=step))
    return

def setup_logging(option_group):
    """Sets up logging in a syslog format by log level
    :param option_group: options as returned by the OptionParser
    """
    stderr_log_format = "%(levelname) -10s %(asctime)s %(funcName) -20s line:%(lineno) -5d: %(message)s"
    file_log_format = "%(asctime)s - %(levelname)s - %(message)s"
    logger = logging.getLogger()
    if option_group.debug:
        logger.setLevel(level=logging.DEBUG)
    elif option_group.verbose:
        logger.setLevel(level=logging.INFO)
    elif option_group.verbose:
        logger.setLevel(level=logging.WARNING)

    handlers = []
    if option_group.syslog:
        handlers.append(logging.SyslogHandler(facility=option_group.syslog))
        # Use standard format here because timestamp and level will be added by syslogd.
    if option_group.logfile:
        handlers.append(logging.FileHandler(option_group.logfile))
        handlers[0].setFormatter(logging.Formatter(file_log_format))
    if not handlers:
        handlers.append(logging.StreamHandler())
        handlers[0].setFormatter(logging.Formatter(stderr_log_format))
    for handler in handlers:
        logger.addHandler(handler)
    return

def repair(options):
    """Repair a keyspace/columnfamily by breaking each token range into $start_steps ranges
    :param options.keyspace: Cassandra keyspace to repair
    :param options.host: (optional) Hostname to pass to nodetool
    :param options.steps: Number of sub-ranges to split primary range in to
    :param options.workers: Number of workers to use
    """
    tokens = Token_Container(options)

    worker_pool = multiprocessing.Pool(options.workers)

    for token_num, host_token in enumerate(tokens.host_tokens):
        range_termination = host_token
        range_start = tokens.get_preceding_token(range_termination)

        if token_num < options.offset:
            logging.info(
                "[{count}/{total}] skipping token..".format(
                    count=token_num + 1,
                    total=tokens.host_token_count))
            continue
        logging.info(
            "[{count}/{total}] repairing range ({token}, {termination}) in {steps} steps for keyspace {keyspace}".format(
                count=token_num + 1,
                total=tokens.host_token_count,
                token=tokens.format(range_start),
                termination=tokens.format(range_termination),
                steps=options.steps,
                keyspace=options.keyspace or "<all>"))

        results = [worker_pool.apply_async(repair_range,
                                           (options,
                                            start,
                                            end,
                                            step,
                                            "{count}/{total}".format(count=token_num + 1,
                                                                     total=tokens.host_token_count)))
                   for start, end, step in tokens.sub_range_generator(range_start, range_termination, options.steps)]
        for r in results:
            r.get()
    return

def main():
    """Validate arguments and initiate repair
    """
    parser = OptionParser()
    parser.add_option("-k", "--keyspace", dest="keyspace", metavar="KEYSPACE",
                      help="Keyspace to repair (REQUIRED)")

    parser.add_option("-c", "--columnfamily", dest="columnfamily", default=[],
                      action="append", metavar="COLUMNFAMILY",
                      help="ColumnFamily to repair, can appear multiple times")

    parser.add_option("-H", "--host", dest="host", default=platform.node(),
                      metavar="HOST", help="Hostname to repair [default: %default]")

    parser.add_option("-s", "--steps", dest="steps", type="int", default=100,
                      metavar="STEPS", help="Number of discrete ranges [default: %default]")

    parser.add_option("-o", "--offset", dest="offset", type="int", default=0,
                      metavar="OFFSET", help="Number of tokens to skip [default: %default]")

    parser.add_option("-n", "--nodetool", dest="nodetool", default="nodetool",
                      metavar="NODETOOL", help="Path to nodetool [default: %default]")

    # The module default for workers is actually the CPU count, but we're
    # going to override it to 1, which matches the old behavior of serial
    # repairs.
    parser.add_option("-w", "--workers", dest="workers", type="int", default=1,
                      metavar="WORKERS", help="Number of workers to use for parallelism [default: %default]")

    parser.add_option("-D", "--datacenter", dest="datacenter", default=None,
                      metavar="DATACENTER", help="Identify local datacenter [default: %default]")

    parser.add_option("-l", "--local", dest="local", default="",
                      action="store_const", const="-local",
                      metavar="LOCAL", help="Restrict repair to the local DC")

    parser.add_option("-p", "--par", dest="par", default="",
                      action="store_const", const="-par",
                      metavar="PAR", help="Carry out a parallel repair (post-2.x only)")

    parser.add_option("-i", "--inc", dest="inc", default="",
                      action="store_const", const="-inc",
                      metavar="INC", help="Carry out an incremental repair (post-2.1 only). Forces --par")

    parser.add_option("-S", "--snapshot", dest="snapshot", default="",
                      action="store_const", const="-snapshot",
                      metavar="LOCAL", help="Use snapshots (pre-2.x only)")

    parser.add_option("-v", "--verbose", dest="verbose", action='store_true',
                      default=False, help="Verbose output")

    parser.add_option("-d", "--debug", dest="debug", action='store_true',
                      default=False, help="Debugging output")

    parser.add_option("--dry-run", dest="dry_run", action='store_true',
                      default=False, help="Do not execute repairs.")

    parser.add_option("--syslog", dest="syslog", metavar="FACILITY",
                      help="Send log messages to the syslog")

    parser.add_option("--logfile", dest="logfile", metavar="FILENAME",
                      help="Send log messages to a file")




    (options, args) = parser.parse_args()

    setup_logging(options)

    if options.columnfamily and not options.keyspace: # keyspace is a *required* for columfamilies
        parser.print_help()
        logging.debug('Invalid configuration options')
        sys.exit(1)

    if args:                    # There are no positional parameters
        parser.print_help()
        logging.debug('Extra parameters')
        sys.exit(1)

    if options.inc and not options.par:
        logging.info('Incremental repairs needs --par: enabling')
        options.par = '-par'

    repair(options)
    exit(0)

if __name__ == "__main__":
    main()
